dask_mongo/tests/test_core.py

```diff
@@ -187,3 +200,81 @@ def test_read_mongo_chunksize(connection_kwargs):
         3,
         1,
     )
+
+
+def test_connection_pooling(connection_kwargs):
+    records = gen_data(size=10)
+    database = "test-db"
+    collection = "test-collection"
+    _close_clients()
+    _cache_inner.cache_clear()
+    mongo_client = _get_client(connection_kwargs)
+    database_ = mongo_client.get_database(database)
+    database_[collection].insert_many(deepcopy(records))
+    for _ in range(3):
+        read_mongo(
+            database,
+            collection,
+            chunksize=5,
+            connection_kwargs=connection_kwargs,
+        )
+    assert _get_num_clients() == 1
+    connection_kwargs.update({"host": ["localhost", "localhost:27017"]})
+    read_mongo(
+        database,
+        collection,
+        chunksize=5,
+        connection_kwargs=connection_kwargs,
+    )
+
+    assert _get_num_clients() == 2
+    for i in range(round(_CACHE_SIZE * 1.2)):
+        connection_kwargs.update({"maxPoolSize": i})
+        read_mongo(
+            database,
+            collection,
+            chunksize=5,
+            connection_kwargs=connection_kwargs,
+        )
+    _close_clients()
+
+    assert _get_num_clients() == _CACHE_SIZE
+    assert len(_CLIENTS) == 0
+
+
+def test_connection_pooling_hashing(connection_kwargs):
+    _close_clients()
+    _cache_inner.cache_clear()
+
+    opts1 = AutoEncryptionOpts(
+        {"local": {"key": b"\x00" * 96}},
+        "k.d",
+        key_vault_client=_get_client(connection_kwargs),
+    )
+    client1 = _get_client(dict(connection_kwargs, **{"auto_encryption_opts": opts1}))
+    client2 = _get_client(
+        dict(
+            connection_kwargs,
+            **{"auto_encryption_opts": opts1, "host": ["localhost", "localhost:27017"]},
+        )
+    )
+    opts3 = AutoEncryptionOpts(
+        {"local": {"key": b"\x00" * 96}},
+        "k.d",
+        key_vault_client=_get_client(connection_kwargs),
+    )
+    client3 = _get_client(dict(connection_kwargs, **{"auto_encryption_opts": opts3}))
+
+    def get_client_opts(client):
+        return client._MongoClient__options._ClientOptions__auto_encryption_opts
```

Comment:
We should not use pymongo internals here. Can we compare the clients using identity instead, ie `assert client is client2`?

Reply:
You should be able to use `client1 is client2` instead of `client1 == client2`.